great_expectations_cloud/agent/agent.py

```diff
@@ -159,10 +160,14 @@ def __init__(self: Self):
 
     def run(self) -> None:
         """Open a connection to GX Cloud."""
+        LOGGER.info("Opening connection to GX Cloud.")
 
-        print("Opening connection to GX Cloud.")
         self._listen()
-        print("The connection to GX Cloud has been closed.")
+        # we need a heavy-handed approach here to kill the main process to prevent a scenario where the agent is
+        # hanging due to a thread that has entered an unhealthy state where it will never finish, for example, in the
+        # case of a RabbitMQ delivery acknowledgement timeout
+        os.kill(os.getpid(), signal.SIGKILL)
```

Comment:
This doesn't feel right. Why/how is the thread getting into a state where it can never finish? Can we do something on that thread to make sure it's healthy? Which thread is getting stuck?

Reply:
Yeah, it does feel gross... but there was a lot of back and forth trying to figure out how to do this between myself, @anthonyburdi and @billdirks. 

For context: https://greatexpectationslabs.slack.com/archives/C05R8RE43CL/p1730269919669839?thread_ts=1730242168.373279&cid=C05R8RE43CL

Happy to discuss more, but what was happening is that the `ThreadPoolExecutor` we define [here](https://github.com/great-expectations/cloud/blob/main/great_expectations_cloud/agent/agent.py#L155) was stuck waiting on its worker thread when a DAT error would get thrown. 

If you have any other ideas about how we can shut down, super interested!

Reply:
I don't think the worker thread is actually in an unhealthy state. I think what is going on is:
1. This thread is the one connected to rabbit (the `_listen` call above). Our rabbit connection goes into a bad state (eg rabbit dies for an upgrade), so we exit the _listen call.
2. The other thread is the worker thread. It is happily working but may be doing a task that takes a long time. If this task were to complete we have no way of acknowledging it to rabbit since that connection is broken.
3. Because of this we want to exit the process. Exiting this thread (eg `sys.exit()`) doesn't exit the whole process. This is because we are using a `ThreadPoolExecutor`. See the [docs here](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor):
>  All threads enqueued to ThreadPoolExecutor will be joined before the interpreter can exit. Note that the exit handler which does this is executed before any exit handlers added using atexit. This means exceptions in the main thread must be caught and handled in order to signal threads to exit gracefully. For this reason, it is recommended that ThreadPoolExecutor not be used for long-running tasks.
4. Since we have no API way of forcing an exit we `kill`.

I didn't understand all these details when we talked yesterday.

Probably the easiest better solution be to avoid the `ThreadPoolExecutor` altogether. Our pool is of size 1 here, we could use the threading API directly. We could make the worker thread a `daemon` thread so it wouldn't keep the process alive if the main thread exited. Even if we spawned a new thread for each task continued running only 1 task at a time, our performance would probably be fine here since that should be fast with respect to job length. 

A harder, but doable, solution would be to implement our own `Executor`, as the docs warn this one isn't meant for long lived tasks. Forcing the call to join to complete before exiting seems like a detractor to this implementation.

Reply:
@billdirks thanks for the clarification.

To keep scope of this fix low, let's keep the `kill` and put more thought into a longer-term solution.

I think we should also consider:
- whether in fact we should be timing out/killing jobs like this. If we find that validation runs are taking over 2 hours, we should figure out how to make them faster or how to restrict the amount of work per single task to make sure they finish
- whether we're certain that we want to do the ACK after the job completes instead of moving it before

Reply:
One other thought (I was talking to @Shinnnyshinshin about this), is maybe we can manage the connection to rabbitmq better. However, we're trying to reconnect now isn't working. I haven't looked at how that is implemented (are we trying to use the same connection object? What are it's semantics?). Maybe we need to reinitiate the connection de novo?

Reply:
We've moved conversation to [this](https://greatexpectations.atlassian.net/wiki/spaces/SUP/pages/1431764993/GX+Agent+Runner+Resilience) doc. It goes into more detail about the issues with reconnecting.